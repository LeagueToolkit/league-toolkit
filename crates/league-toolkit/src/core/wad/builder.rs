use std::io::{self, BufReader, BufWriter, Cursor, Read, Seek, SeekFrom, Write};

use byteorder::{WriteBytesExt, LE};
use flate2::read::GzEncoder;
use io_ext::measure;
use itertools::Itertools;
use xxhash_rust::{xxh3, xxh64};

use crate::league_file::LeagueFileKind;

use super::{WadChunk, WadChunkCompression, WadError};

#[derive(Debug, thiserror::Error)]
pub enum WadBuilderError {
    #[error("wad error")]
    WadError(#[from] WadError),

    #[error("io error")]
    IoError(#[from] io::Error),

    #[error("unsupported compression type: {0}")]
    UnsupportedCompressionType(WadChunkCompression),
}

/// Implements a builder interface for creating WAD files.
///
/// # Examples
/// ```
/// # use league_toolkit::core::wad::*;
/// #
/// let builder = WadBuilder::default();
/// builder.with_chunk(WadChunkBuilder::default().with_path("path/to/chunk"));
/// builder.build_to_writer(File::create("output.wad").unwrap());
/// ```
#[derive(Debug, Default)]
pub struct WadBuilder {
    chunk_builders: Vec<WadChunkBuilder>,
}

impl WadBuilder {
    pub fn with_chunk(mut self, chunk: WadChunkBuilder) -> Self {
        self.chunk_builders.push(chunk);
        self
    }

    /// Build the WAD file and write it to the given writer.
    ///
    /// * `writer` - The writer to write the WAD file to.
    /// * `provide_chunk_data` - A function that provides the rawdata for each chunk.
    pub fn build_to_writer<
        TWriter: io::Write + io::Seek,
        TChunkDataProvider: Fn(u64, &mut Cursor<&mut Vec<u8>>) -> Result<(), WadBuilderError>,
    >(
        self,
        writer: TWriter,
        provide_chunk_data: TChunkDataProvider,
    ) -> Result<(), WadBuilderError> {
        // First we need to write a dummy header and TOC, so we can calculate from where to start writing the chunks
        let mut writer = BufWriter::new(writer);

        let (header_toc_size, toc_offset) = self.write_dummy_toc::<TWriter>(&mut writer)?;

        let ordered_chunks = self
            .chunk_builders
            .iter()
            .sorted_by_key(|chunk| chunk.path)
            .collect::<Vec<_>>();

        let mut final_chunks = Vec::new();

        let mut current_data_offset = toc_offset + header_toc_size;
        for chunk in ordered_chunks {
            let mut chunk_data = Vec::new();
            provide_chunk_data(chunk.path, &mut Cursor::new(&mut chunk_data))?;

            let chunk_data_size = chunk_data.len();
            let compressed_data = Self::compress_chunk_data(&chunk_data, chunk.force_compression)?;
            let compressed_data_size = compressed_data.len();
            let compressed_checksum = xxh3::xxh3_64(&compressed_data);

            writer.write_all(&compressed_data)?;

            current_data_offset = current_data_offset.wrapping_add(compressed_data_size as u64);

            final_chunks.push(WadChunk {
                path_hash: chunk.path,
                data_offset: current_data_offset as usize,
                compressed_size: compressed_data_size,
                uncompressed_size: chunk_data_size,
                compression_type: WadChunkCompression::Zstd,
                is_duplicated: false,
                frame_count: 0,
                start_frame: 0,
                checksum: compressed_checksum,
            });
        }

        writer.seek(SeekFrom::Start(toc_offset))?;

        for chunk in final_chunks {
            chunk.write_v3_4(&mut writer)?;
        }

        Ok(())
    }

    fn write_dummy_toc<W: io::Write + io::Seek>(
        &self,
        writer: &mut BufWriter<W>,
    ) -> Result<(u64, u64), WadBuilderError> {
        let (header_toc_size, toc_offset) = measure(writer, |writer| {
            // Write the header
            writer.write_u16::<LE>(0x5752)?;
            writer.write_u8(3)?; // major
            writer.write_u8(4)?; // minor

            // Write dummy ECDSA signature
            writer.write_all(&[0; 256])?;
            writer.write_u64::<LE>(0)?;

            // Write dummy TOC
            let toc_offset = writer.stream_position()?;
            for _ in self.chunk_builders.iter() {
                writer.write_all(&[0; 32])?;
            }

            Ok::<_, WadBuilderError>(toc_offset)
        })?;

        Ok((header_toc_size, toc_offset))
    }

    fn compress_chunk_data(
        data: &[u8],
        force_compression: Option<WadChunkCompression>,
    ) -> Result<Vec<u8>, WadBuilderError> {
        let compressed_data = match force_compression {
            Some(compression) => Self::compress_chunk_data_by_compression(data, compression)?,
            None => {
                let kind = LeagueFileKind::identify_from_bytes(data);

                Self::compress_chunk_data_by_compression(data, kind.ideal_compression())?
            }
        };

        Ok(compressed_data)
    }

    fn compress_chunk_data_by_compression(
        data: &[u8],
        compression: WadChunkCompression,
    ) -> Result<Vec<u8>, WadBuilderError> {
        let mut compressed_data = Vec::new();
        match compression {
            WadChunkCompression::None => {
                compressed_data = data.to_vec();
            }
            WadChunkCompression::GZip => {
                let reader = BufReader::new(data);
                let mut encoder = GzEncoder::new(reader, flate2::Compression::default());

                encoder.read_to_end(&mut compressed_data)?;
            }
            WadChunkCompression::Zstd => {
                zstd::Encoder::new(BufWriter::new(&mut compressed_data), 3)?.write_all(data)?
            }
            WadChunkCompression::Satellite => {
                return Err(WadBuilderError::UnsupportedCompressionType(compression));
            }
            WadChunkCompression::ZstdMulti => {
                return Err(WadBuilderError::UnsupportedCompressionType(compression));
            }
        }

        Ok(compressed_data)
    }
}

#[derive(Debug, Clone, Copy, Default)]
pub struct WadChunkBuilder {
    path: u64,

    /// If provided, the chunk will be compressed using the given compression type, otherwise the ideal compression will be used.
    force_compression: Option<WadChunkCompression>,
}

impl WadChunkBuilder {
    pub fn with_path(mut self, path: impl AsRef<str>) -> Self {
        self.path = xxh64::xxh64(path.as_ref().to_lowercase().as_bytes(), 0);
        self
    }

    pub fn with_force_compression(mut self, compression: WadChunkCompression) -> Self {
        self.force_compression = Some(compression);
        self
    }
}
