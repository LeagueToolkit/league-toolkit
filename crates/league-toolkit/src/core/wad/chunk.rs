use std::io::{BufReader, Read};

use byteorder::{ReadBytesExt as _, LE};
use num_enum::{IntoPrimitive, TryFromPrimitive};

use super::WadError;

#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
#[repr(u8)]
pub enum WadChunkCompression {
    None = 0,
    GZip = 1,
    Satellite = 2,
    Zstd = 3,
    ZstdMulti = 4,
}

#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct WadChunk {
    pub path_hash: u64,
    pub data_offset: usize,
    pub compressed_size: usize,
    pub uncompressed_size: usize,
    pub compression_type: WadChunkCompression,
    pub is_duplicated: bool,
    pub frame_count: u8,
    pub start_frame: u16,
    pub checksum: u64,
}

impl WadChunk {
    pub(crate) fn read<R: Read>(reader: &mut BufReader<R>) -> Result<WadChunk, WadError> {
        let path_hash = reader.read_u64::<LE>()?;
        let data_offset = reader.read_u32::<LE>()? as usize;
        let compressed_size = reader.read_i32::<LE>()? as usize;
        let uncompressed_size = reader.read_i32::<LE>()? as usize;

        let type_frame_count = reader.read_u8()?;
        let frame_count = type_frame_count >> 4;
        let compression_type = WadChunkCompression::try_from_primitive(type_frame_count & 0xF)
            .expect("failed to read chunk compression");

        let is_duplicated = reader.read_u8()? == 1;
        let start_frame = reader.read_u16::<LE>()?;
        let checksum = reader.read_u64::<LE>()?;

        Ok(WadChunk {
            path_hash,
            data_offset,
            compressed_size,
            uncompressed_size,
            compression_type,
            is_duplicated,
            frame_count,
            start_frame,
            checksum,
        })
    }

    pub fn path_hash(&self) -> u64 {
        self.path_hash
    }
    pub fn data_offset(&self) -> usize {
        self.data_offset
    }
    pub fn compressed_size(&self) -> usize {
        self.compressed_size
    }
    pub fn uncompressed_size(&self) -> usize {
        self.uncompressed_size
    }
    pub fn compression_type(&self) -> WadChunkCompression {
        self.compression_type
    }
    pub fn checksum(&self) -> u64 {
        self.checksum
    }
}
