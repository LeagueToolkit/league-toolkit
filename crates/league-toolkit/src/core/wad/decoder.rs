use std::io::{Read, Seek, SeekFrom};

use super::{WadChunk, WadChunkCompression, WadError};
use flate2::read::GzDecoder;
use memchr::memmem;

const ZSTD_MAGIC: [u8; 4] = [0x28, 0xB5, 0x2F, 0xFD];

#[cfg(all(feature = "zstd", feature = "ruzstd"))]
compile_error!("feature \"zstd\" and feature \"ruzstd\" cannot be enabled at the same time");

pub struct WadDecoder<'wad, TSource: Read + Seek> {
    pub(crate) source: &'wad mut TSource,
}

impl<'wad, TSource> WadDecoder<'wad, TSource>
where
    TSource: Read + Seek,
{
    pub fn load_chunk_raw(&mut self, chunk: &WadChunk) -> Result<Box<[u8]>, WadError> {
        let mut data = vec![0; chunk.compressed_size];

        self.source
            .seek(SeekFrom::Start(chunk.data_offset as u64))?;
        self.source.read_exact(&mut data)?;

        Ok(data.into_boxed_slice())
    }
    pub fn load_chunk_decompressed(&mut self, chunk: &WadChunk) -> Result<Box<[u8]>, WadError> {
        match chunk.compression_type {
            WadChunkCompression::None => self.load_chunk_raw(chunk),
            WadChunkCompression::GZip => self.decode_gzip_chunk(chunk),
            WadChunkCompression::Satellite => Err(WadError::Other(String::from(
                "satellite chunks are not supported",
            ))),
            WadChunkCompression::Zstd => self.decode_zstd_chunk(chunk),
            WadChunkCompression::ZstdMulti => self.decode_zstd_multi_chunk(chunk),
            _ => todo!(),
        }
    }

    fn decode_gzip_chunk(&mut self, chunk: &WadChunk) -> Result<Box<[u8]>, WadError> {
        self.source
            .seek(SeekFrom::Start(chunk.data_offset as u64))?;

        let mut data = vec![0; chunk.uncompressed_size];
        log::debug!("decoding gzip chunk...");
        GzDecoder::new(&mut self.source).read_exact(&mut data)?;

        Ok(data.into_boxed_slice())
    }
    fn decode_zstd_chunk(&mut self, chunk: &WadChunk) -> Result<Box<[u8]>, WadError> {
        self.source
            .seek(SeekFrom::Start(chunk.data_offset as u64))?;

        let mut data: Vec<u8> = vec![0; chunk.uncompressed_size];

        log::debug!("decoding zstd chunk...");
        #[cfg(feature = "zstd")]
        {
            zstd::Decoder::new(&mut self.source)
                .expect("failed to create zstd decoder")
                .read_exact(&mut data)?;
        }
        #[cfg(feature = "ruzstd")]
        {
            ruzstd::StreamingDecoder::new(&mut self.source)
                .expect("failed to create ruzstd decoder")
                .read_exact(&mut data)?;
        }

        Ok(data.into_boxed_slice())
    }
    fn decode_zstd_multi_chunk(&mut self, chunk: &WadChunk) -> Result<Box<[u8]>, WadError> {
        let raw_data = self.load_chunk_raw(chunk)?;
        let mut data: Vec<u8> = vec![0; chunk.uncompressed_size];

        let zstd_magic_offset =
            memmem::find(&raw_data, &ZSTD_MAGIC).ok_or(WadError::DecompressionFailure {
                path_hash: chunk.path_hash,
                reason: String::from("failed to find zstd magic"),
            })?;

        // copy raw uncompressed data which exists before first zstd frame
        for (i, value) in raw_data[0..zstd_magic_offset].iter().enumerate() {
            data[i] = *value;
        }

        // seek to start of first zstd frame
        self.source.seek(SeekFrom::Start(
            (chunk.data_offset + zstd_magic_offset) as u64,
        ))?;

        log::debug!(
            "decoding zstd multi chunk...\ndata_off: {}\nmagic_off: {zstd_magic_offset}",
            chunk.data_offset
        );
        // decode zstd data
        #[cfg(feature = "zstd")]
        {
            zstd::Decoder::new(&mut self.source)
                .expect("failed to create zstd decoder")
                .read_exact(&mut data[zstd_magic_offset..])?;
        }
        #[cfg(feature = "ruzstd")]
        {
            ruzstd::StreamingDecoder::new(&mut self.source)
                .expect("failed to create ruzstd decoder")
                .read(&mut data[zstd_magic_offset..])?;
        }

        Ok(data.into_boxed_slice())
    }
}
