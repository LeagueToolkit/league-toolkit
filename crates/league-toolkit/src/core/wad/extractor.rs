use std::{
    fs::File,
    io::{Read, Seek, Write},
    path::{Path, PathBuf},
};

use crate::{
    core::wad::{WadHashtableExt, WadHashtablePath},
    league_file::LeagueFileKind,
};

use super::{Wad, WadChunk, WadDecoder, WadError, WadHashtable};

impl<TSource: Read + Seek> Wad<TSource> {
    /// Extract all chunks from the WAD file to the given directory.
    ///
    /// # Arguments
    ///
    /// * `output_dir` - The directory to extract the chunks to.
    /// * `hashtable` - The hashtable to use to resolve the chunk paths.
    pub fn extract_all(
        &mut self,
        output_dir: impl AsRef<Path>,
        hashtable: &WadHashtable,
    ) -> Result<(), WadError> {
        let (mut decoder, chunks) = self.decode();
        for chunk in chunks.values() {
            Self::extract_chunk(chunk, &output_dir, &mut decoder, hashtable)?;
        }

        Ok(())
    }

    /// Extract a chunk from the WAD file to the given directory.
    pub fn extract_chunk(
        chunk: &WadChunk,
        output_dir: impl AsRef<Path>,
        decoder: &mut WadDecoder<TSource>,
        hashtable: &WadHashtable,
    ) -> Result<(), WadError> {
        let data = decoder.load_chunk_decompressed(chunk)?;

        let chunk_path = hashtable.resolve_or_default(chunk.path_hash);

        let output_path = match chunk_path {
            WadHashtablePath::Default(path) => {
                let path = Path::new(path);
                let dir = output_dir.as_ref().join(path.parent().unwrap());
                std::fs::create_dir_all(&dir)?;

                output_dir.as_ref().join(path)
            }
            WadHashtablePath::Unknown(name) => {
                let name = Self::resolve_unknown_chunk_name(&name, &data);

                output_dir.as_ref().join(&name)
            }
        };

        let mut file = File::create(output_path)?;
        file.write_all(&data)?;

        Ok(())
    }

    fn resolve_unknown_chunk_name(name: &str, data: &[u8]) -> PathBuf {
        let path = Path::new(name);
        let extension = path.extension().unwrap_or_default();

        let kind = match LeagueFileKind::from_extension(extension.to_str().unwrap_or_default()) {
            LeagueFileKind::Unknown => LeagueFileKind::identify_from_bytes(data),
            kind => kind,
        };

        match kind {
            // If the final kind is unknown, prefix with dot
            LeagueFileKind::Unknown => {
                let mut path = PathBuf::new();
                path.push(".");
                path.push(name);
                path
            }
            _ => path.with_extension(kind.extension().unwrap_or_default()),
        }
    }
}
