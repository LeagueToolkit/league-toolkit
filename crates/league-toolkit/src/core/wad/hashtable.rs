use std::{
    collections::HashMap,
    io::{self, BufRead, BufReader},
    num::ParseIntError,
};

#[derive(Debug, thiserror::Error)]
pub enum WadHashtableError {
    #[error("Invalid hash: {0}")]
    InvalidHash(#[from] ParseIntError),

    #[error(transparent)]
    Io(#[from] io::Error),
}

pub type WadHashtable = HashMap<u64, String>;

pub enum WadHashtableHash<'a> {
    Default(&'a str),
    Unknown(String),
}

pub trait WadHashtableExt {
    fn from_reader(reader: impl io::Read) -> Result<WadHashtable, WadHashtableError>;

    fn resolve(&self, hash: u64) -> Option<&str>;
    fn resolve_or_default(&self, hash: u64) -> WadHashtableHash;
}

impl WadHashtableExt for WadHashtable {
    fn from_reader(reader: impl io::Read) -> Result<Self, WadHashtableError> {
        let reader = BufReader::new(reader);
        let mut hashtable = HashMap::new();

        for line in reader.lines() {
            let line = line?;

            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 {
                let hash = parts[0].trim();
                let path = parts[1].trim();
                hashtable.insert(u64::from_str_radix(hash, 16)?, path.to_string());
            }
        }

        Ok(hashtable)
    }

    fn resolve(&self, hash: u64) -> Option<&str> {
        match self.get(&hash) {
            Some(path) => Some(path.as_str()),
            None => None,
        }
    }

    fn resolve_or_default(&self, hash: u64) -> WadHashtableHash {
        match self.resolve(hash) {
            Some(path) => WadHashtableHash::Default(path),
            None => WadHashtableHash::Unknown(format!("0x{:x}", hash)),
        }
    }
}
