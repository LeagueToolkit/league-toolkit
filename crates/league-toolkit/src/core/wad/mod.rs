//! Wad file handling

mod builder;
mod chunk;
mod decoder;
mod error;

pub use builder::*;
pub use chunk::*;
pub use decoder::*;
pub use error::*;

use std::{
    collections::HashMap,
    io::{BufReader, Read, Seek, SeekFrom},
};

use byteorder::{ReadBytesExt as _, LE};

#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug)]
/// A wad file
pub struct Wad<TSource: Read + Seek> {
    chunks: HashMap<u64, WadChunk>,
    #[cfg_attr(feature = "serde", serde(skip))]
    source: TSource,
}

impl<TSource: Read + Seek> Wad<TSource> {
    pub fn chunks(&self) -> &HashMap<u64, WadChunk> {
        &self.chunks
    }

    pub fn mount(mut source: TSource) -> Result<Wad<TSource>, WadError> {
        let mut reader = BufReader::new(&mut source);

        // 0x5752 = "RW"
        let magic = reader.read_u16::<LE>()?;
        if magic != 0x5752 {
            return Err(WadError::InvalidHeader {
                expected: String::from("RW"),
                actual: format!("0x{:x}", magic),
            });
        }

        let major = reader.read_u8()?;
        let minor = reader.read_u8()?;
        if major > 3 {
            return Err(WadError::InvalidVersion { major, minor });
        }

        if major == 2 {
            let _ecdsa_length = reader.seek(SeekFrom::Current(1))?;
            let _ecdsa_signature = reader.seek(SeekFrom::Current(83))?;
            let _data_checksum = reader.seek(SeekFrom::Current(8))?;
        } else if major == 3 {
            let _ecdsa_signature = reader.seek(SeekFrom::Current(256))?;
            let _data_checksum = reader.seek(SeekFrom::Current(8))?;
        }

        if major == 1 || major == 2 {
            let _toc_start_offset = reader.seek(SeekFrom::Current(2))?;
            let _toc_chunk_size = reader.seek(SeekFrom::Current(2))?;
        }

        let chunk_count = reader.read_i32::<LE>()? as usize;
        let mut chunks = HashMap::<u64, WadChunk>::with_capacity(chunk_count);
        for _ in 0..chunk_count {
            let chunk = WadChunk::read_v3_1(&mut reader)?;
            chunks
                .insert(chunk.path_hash(), chunk)
                .map_or(Ok(()), |chunk| {
                    Err(WadError::DuplicateChunk {
                        path_hash: chunk.path_hash(),
                    })
                })?;
        }

        Ok(Wad { chunks, source })
    }

    pub fn decode(&mut self) -> (WadDecoder<'_, TSource>, &HashMap<u64, WadChunk>) {
        (
            WadDecoder {
                source: &mut self.source,
            },
            &self.chunks,
        )
    }
}
