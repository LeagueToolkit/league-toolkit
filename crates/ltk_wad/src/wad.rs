use std::collections::BTreeMap;

use binrw::{binrw, BinRead, BinWrite, NamedArgs};
use derive_more::{AsMut, AsRef, Debug, Deref, DerefMut, From, Index, Into};
use itertools::Itertools;

use crate::{
    entry::{self, EntryExt},
    header::{Header, HeaderExt},
};

#[binrw]
#[brw(magic = b"RW", little)]
#[derive(Debug, Clone)]
pub struct Wad<E = EntriesMap>
where
    for<'a> E: BinRead<Args<'a> = EntriesArgs> + BinWrite<Args<'a> = EntriesArgs>,
{
    pub header: Header,
    #[brw(args{major: header.major(), minor: header.minor(), count: header.entry_count().try_into().unwrap()})]
    pub entries: E,
}

#[binrw]
#[brw(import_raw(args: EntriesArgs))]
#[derive(AsMut, AsRef, Debug, Deref, DerefMut, Index, Into, From)]
pub struct EntriesMap(
    #[br(count = args.count)]
    #[br(args {inner: binrw::args!{major: args.major, minor: args.minor}})]
    #[bw(args {major: args.major, minor: args.minor})]
    #[bw(map = |entries| entries.values().copied().collect::<Vec<entry::Entry>>())]
    #[br(map = |entries: Vec<entry::Entry>| entries.into_iter().map(|e| (e.path_hash(), e)).collect() ) ]
    pub BTreeMap<u64, entry::Entry>,
);

#[derive(Clone, NamedArgs)]
pub struct EntriesArgs {
    pub major: u8,
    pub minor: u8,
    pub count: usize,
}

type EntriesDyn<'a> = Box<dyn Iterator<Item = &'a dyn EntryExt> + 'a>;

impl<E> Wad<E>
where
    for<'a> E: BinRead<Args<'a> = EntriesArgs> + BinWrite<Args<'a> = EntriesArgs>,
{
    pub fn major(&self) -> u8 {
        self.header.major()
    }

    pub fn minor(&self) -> u8 {
        self.header.minor()
    }

    pub fn version(&self) -> (u8, u8) {
        (self.major(), self.minor())
    }
}

impl Wad<Entries> {
    pub fn entries_dyn(&self) -> EntriesDyn<'_> {
        match &self.entries {
            Entries::V3_4(btree_map) => Box::new(btree_map.values().map(|v| v as &dyn EntryExt)),
            Entries::V3(btree_map) => Box::new(btree_map.values().map(|v| v as &dyn EntryExt)),
            Entries::V2(btree_map) => Box::new(btree_map.values().map(|v| v as &dyn EntryExt)),
            Entries::V1(btree_map) => Box::new(btree_map.values().map(|v| v as &dyn EntryExt)),
        }
    }
}

#[derive(BinRead, BinWrite)]
#[brw(import_raw(args: EntriesArgs))]
#[derive(Clone, Debug)]
pub enum Entries {
    #[br(pre_assert(args.major == 3 && args.minor == 4))]
    V3_4(
        #[br(count = args.count, map = vec_to_btree::<entry::V3_4>)]
        #[bw(map = |e| e.values().copied().collect_vec())]
        BTreeMap<u64, entry::V3_4>,
    ),
    #[br(pre_assert(args.major == 3 && args.minor < 4))]
    V3(
        #[br(count = args.count, map = vec_to_btree::<entry::V3>)]
        #[bw(map = |e| e.values().copied().collect_vec())]
        BTreeMap<u64, entry::V3>,
    ),
    #[br(pre_assert(args.major == 2))]
    V2(
        #[br(count = args.count, map = vec_to_btree::<entry::V2>)]
        #[bw(map = |e| e.values().copied().collect_vec())]
        BTreeMap<u64, entry::V2>,
    ),
    #[br(pre_assert(args.major == 1))]
    V1(
        #[br(count = args.count, map = vec_to_btree::<entry::V1>)]
        #[bw(map = |e| e.values().copied().collect_vec())]
        BTreeMap<u64, entry::V1>,
    ),
}

fn vec_to_btree<T: EntryExt>(entries: Vec<T>) -> BTreeMap<u64, T> {
    entries
        .into_iter()
        .map(|entry| (entry.path_hash(), entry))
        .collect()
}
