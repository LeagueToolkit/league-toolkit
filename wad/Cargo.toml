[package]
name = "wad"
version = "0.1.0"
edition = "2021"

[dependencies]
byteorder = "1.5.0"
flate2 = "1.0.30"
memchr = "2.7.2"
num_enum = "0.7.2"
thiserror = "1.0.60"
zstd = "0.13.1"
