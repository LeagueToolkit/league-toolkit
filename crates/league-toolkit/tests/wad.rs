use insta::assert_ron_snapshot;
use league_toolkit::core::wad::Wad;
use std::io::Cursor;

#[test]
fn read() {
    let r = Cursor::new(include_bytes!("wads/UI.en_US.wad.client.v3-4"));
    let wad = Wad::mount(r).unwrap();

    insta::with_settings!({sort_maps => true}, {
        assert_ron_snapshot!(wad.chunks());
    });
}
