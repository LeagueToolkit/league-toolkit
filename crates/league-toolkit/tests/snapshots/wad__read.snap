---
source: crates/league-toolkit/tests/wad.rs
expression: wad.chunks()
---
{
  2227312568830940929: WadChunk(
    path_hash: 2227312568830940929,
    data_offset: 10911,
    compressed_size: 23023,
    uncompressed_size: 131084,
    compression_type: Zstd,
    is_duplicated: false,
    frame_count: 0,
    start_frame: 0,
    checksum: 3922618146135880204,
  ),
  3349600290122526159: WadChunk(
    path_hash: 3349600290122526159,
    data_offset: 33934,
    compressed_size: 20786,
    uncompressed_size: 131084,
    compression_type: Zstd,
    is_duplicated: false,
    frame_count: 0,
    start_frame: 0,
    checksum: 13212711047613655776,
  ),
  11042693609879973350: WadChunk(
    path_hash: 11042693609879973350,
    data_offset: 368,
    compressed_size: 10543,
    uncompressed_size: 32780,
    compression_type: Zstd,
    is_duplicated: false,
    frame_count: 0,
    start_frame: 0,
    checksum: 3841611443327106610,
  ),
}
